### PR TITLE
Convert AS to ES Modules

### PR DESCRIPTION
**Pull request by k88hudson**
In this patch I've converted all js files in the `test/` and `content-src/` subdirectories to use ES modules. E.g.

Old style:
```js
const {Foo: Floo} = require("./Foo");
class Bar {...}
module.exports = Bar;
```
New style:
```js
import {Foo as Floo} from "./Foo";
export class Bar {...}
```
The migration also includes the following changes:

#### eslint checks import sort order and prevents common js
The `sort-imports` rule requires that imports are sorted alphabetically. I also added the `no-commonjs` rule to the relevant directories that prevents the use of `require`, `module.exports`, or `exports`.

#### named exports instead of module.exports = MyComponent
 I changed the style of exporting to use named exports rather than `default`. This results in a slightly different structure for exporting a react component and the unconnected version of itself:

Old style:
```js
class TopSites { ... }
module.exports = connect(connector)(TopSites);
module.exports._unconnected = TopSites;

// importing
const TopSites = require("TopSites.jsx");
const {_unconnected} = require("TopSites.jsx");
```

New style:
```js
export class _TopSites { ... }
export const TopSites = connect(connector)(_TopSites);

// importing
import {TopSites, _TopSites} from "TopSites.jsx";
```

### bundle stats / perf with scope hoisting
So far I'm seeing the following differences in output (about 20kb) with scope hoisting enabled. Unfortunately, no meaninful differences in js executation time on load with profiling.

Here's the [diff of the built file](https://gist.github.com/k88hudson/c62c66e6b6196ffac5b301f635d3d750).

Here are stats:

Without scope hoisting: 
```
Time: 1047ms
activity-stream.bundle.js  177 kB       0  [emitted]  main
```
With scope hoisting including bailout info:
```
Time: 1162ms
                    Asset    Size  Chunks             Chunk Names
activity-stream.bundle.js  156 kB       0  [emitted]  main
   [0] ./system-addon/common/Actions.jsm 7.82 kB {0} [built]
       ModuleConcatenation bailout: Module is not an ECMAScript module
   [4] (webpack)/buildin/global.js 488 bytes {0} [built]
       ModuleConcatenation bailout: Module is not an ECMAScript module
   [5] ./system-addon/common/Reducers.jsm 11.4 kB {0} [built]
       ModuleConcatenation bailout: Module is not an ECMAScript module
   [6] ./system-addon/content-src/components/LinkMenu/LinkMenu.jsx + 2 modules 7.93 kB {0} [built]
       ModuleConcatenation bailout: Cannot concat with ./system-addon/common/Actions.jsm (<- Module is not an ECMAScript module)
       ModuleConcatenation bailout: Cannot concat with external "React" (<- Module is not an ECMAScript module)
       ModuleConcatenation bailout: Cannot concat with external "ReactIntl" (<- Module is not an ECMAScript module)
   [9] ./system-addon/common/PerfService.jsm 4.09 kB {0} [built]
       ModuleConcatenation bailout: Module is not an ECMAScript module
  [10] ./system-addon/content-src/activity-stream.jsx 1.27 kB {0} [built]
       ModuleConcatenation bailout: Module uses injected variables (global)
  [11] ./system-addon/content-src/components/Base/Base.jsx + 9 modules 43.6 kB {0} [built]
       ModuleConcatenation bailout: Cannot concat with ./system-addon/content-src/components/Sections/Sections.jsx (<- Module uses injected variables (global))
       ModuleConcatenation bailout: Cannot concat with ./system-addon/common/Actions.jsm (<- Module is not an ECMAScript module)
       ModuleConcatenation bailout: Cannot concat with ./system-addon/common/PrerenderData.jsm (<- Module is not an ECMAScript module)
       ModuleConcatenation bailout: Cannot concat with external "React" (<- Module is not an ECMAScript module)
       ModuleConcatenation bailout: Cannot concat with external "ReactIntl" (<- Module is not an ECMAScript module)
       ModuleConcatenation bailout: Cannot concat with external "ReactRedux" (<- Module is not an ECMAScript module)
  [12] ./system-addon/common/Dedupe.jsm 935 bytes {0} [built]
       ModuleConcatenation bailout: Module is not an ECMAScript module
  [13] ./system-addon/common/PrerenderData.jsm 2.67 kB {0} [built]
       ModuleConcatenation bailout: Module is not an ECMAScript module
  [14] ./system-addon/content-src/lib/constants.js 91 bytes {0} [built]
       ModuleConcatenation bailout: Module uses injected variables (global)
  [16] ./system-addon/content-src/components/Card/Card.jsx + 1 modules 7.61 kB {0} [built]
       ModuleConcatenation bailout: Cannot concat with ./system-addon/content-src/components/LinkMenu/LinkMenu.jsx because of ./system-addon/content-src/activity-stream.jsx
       ModuleConcatenation bailout: Cannot concat with ./system-addon/common/Actions.jsm (<- Module is not an ECMAScript module)
       ModuleConcatenation bailout: Cannot concat with external "React" (<- Module is not an ECMAScript module)
       ModuleConcatenation bailout: Cannot concat with external "ReactIntl" (<- Module is not an ECMAScript module)
  [18] ./system-addon/content-src/lib/detect-user-session-start.js 2.42 kB {0} [built]
       ModuleConcatenation bailout: Module uses injected variables (global)
  [20] ./system-addon/content-src/lib/snippets.js 11.5 kB {0} [built]
       ModuleConcatenation bailout: Module uses injected variables (global)
  [21] ./system-addon/content-src/lib/init-store.js 4.8 kB {0} [built]
       ModuleConcatenation bailout: Module uses injected variables (global)
    + 9 hidden modules
```

### CODE REVIEW

File: system-addon/test/unit/utils.js [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
-const {mount, shallow} = require("enzyme");
-const {IntlProvider, intlShape} = require("react-intl");
+import {IntlProvider, intlShape} from "react-intl";
+import {mount, shallow} from "enzyme";
+import React from "react";
+
 const messages = require("data/locales.json")["en-US"];
-const intlProvider = new IntlProvider({locale: "en-US", messages});
+
+const intlProvider = new IntlProvider({locale: "en", messages});
```
Comment (Mardak) [inline system-addon/test/unit/utils.js]: What required this "en-US" -> "en" change?

**Comment by k88hudson** [comment]
Ok, this should be everything in `test/` and `content-src/`. I have left the `sort-imports` rule as is and added `no-commonjs` to those directories.

File: system-addon/test/unit/lib/FilterAdult.test.js [hunk 2]
```diff
@@ -14,7 +15,6 @@ describe("filterAdult", () => {
         return hashStub;
       }
     };
-    filterAdult = require("lib/FilterAdult.jsm").filterAdult;
```
Comment (k88hudson) [inline system-addon/test/unit/lib/FilterAdult.test.js]: Is there a reason this is re-imported/re-assigned once per test @Mardak?

Comment (Mardak) [reply system-addon/test/unit/lib/FilterAdult.test.js]: Not that I can remember. A quick glance maybe points to the lazy `gCryptoHash = Cc["@mozilla.org/security/hash;1"].createInstance(Ci.nsICryptoHash);` ? I believe at one point it used some `XPCOM.defineLazy…`something, but I suppose if the tests pass, then no problem!

**Comment by Mardak** [comment]
To be clear, we don't need to change the rule now, and the PR is fine. If I get around to fixing up `sort-imports`, I'll probably treat single and multiple the same time as turning on ignoreCase.

**Comment by Mardak** [comment]
@k88hudson curious, does this look better with `sort-imports`? https://github.com/Mardak/activity-stream/commit/f55282b2b0292eb4940eb35ce90c1ec70a8a6ea2

File: .eslintrc.js [hunk 3]
```diff
@@ -19,7 +20,8 @@ module.exports = {
   "plugins": [
     "json", // require("eslint-plugin-json")
     "promise", // require("eslint-plugin-promise")
-    "react" // require("eslint-plugin-react")
+    "react", // require("eslint-plugin-react")
+    "import" // require("eslint-plugin-import")
```
Comment (Mardak) [inline .eslintrc.js]: nit: "import" before "json" sorted ;)

File: system-addon/content-src/activity-stream-prerender.jsx [hunk 4]
```diff
@@ -1,25 +1,25 @@
...
-const {actionTypes: at} = require("common/Actions.jsm");
-const {PrerenderData} = require("common/PrerenderData.jsm");
+import {INITIAL_STATE, reducers} from "common/Reducers.jsm";
+import {Base} from "content-src/components/Base/Base";
+import {PrerenderData} from "common/PrerenderData.jsm";
+import {Provider} from "react-redux";
+import React from "react";
+import ReactDOM from "react-dom/server";
+import {actionTypes as at} from "common/Actions.jsm";
+import {initStore} from "content-src/lib/init-store";
```
Comment (Mardak) [inline system-addon/content-src/activity-stream-prerender.jsx]: We'll go with `ignoreCase` to have fewer lines changed down the lines.

File: system-addon/content-src/components/ConfirmDialog/ConfirmDialog.jsx [hunk 5]
```diff
@@ -1,7 +1,7 @@
-const React = require("react");
-const {connect} = require("react-redux");
-const {FormattedMessage} = require("react-intl");
-const {actionTypes, actionCreators: ac} = require("common/Actions.jsm");
+import {actionCreators as ac, actionTypes} from "common/Actions.jsm";
```
Comment (pdehaan) [inline system-addon/content-src/components/ConfirmDialog/ConfirmDialog.jsx]: Mildly irrelevant, but I think this was the only place where we import `actionTypes` but don't assign it to an `at` alias. 🤷‍♀️

Comment (k88hudson) [reply system-addon/content-src/components/ConfirmDialog/ConfirmDialog.jsx]: Noted, but I didn't really feel like making the change was worth it in this patch

**Comment by as-pine-proxy** [comment]
[Treeherder: 3a234861f5f9a10ca8e2c5f903dc28ad6a227fb8](https://treeherder.mozilla.org/#/jobs?repo=pine&revision=3a234861f5f9a10ca8e2c5f903dc28ad6a227fb8)

**Pull request merged**